### DISCUSSION TITLE
floating point math to read internal temperature sensor

### DISCUSSION THREAD

**Discussion by marklacas** [tabemann/zeptoforth#33] [Q&A] [answered]
I've figured out how to convert a number to a float s>f and how to divide and multiply, f/ f*.  How does one subtract and add two floating point numbers?

Trying to get the onboard temperature sensor working.

conversion_factor = 3.3 / 65535
reading = adc * conversion_factor
temperature = 27 - (reading - 0.706) / 0.001721

Do I use d+ and d-?

**Comment by tabemann** [comment, marked as answer]
To be exact, zeptoforth does not use floating-point numbers but rather fixed-point numbers, which are stored as double cells due to being in S31.32 format. For multiplication and division you use `f*` and `f/`. However, as addition and subtraction of S31.32 fixed-point numbers is identical to that of normal double-cell integers, you do use `d+` and  `d-` for that.

**Reply by marklacas** [reply]
Thanks!

**Comment by marklacas** [comment]
Here is the final code to read the temperature in C and F

adc import

3,3 4096,0 f/ 2constant conversion_factor

\ Send the CPU temperature out to the console
: temp ( -- )
    temp-adc-chan default-adc adc@ s>f conversion_factor f*
    0,706 d- 
    2swap 0,001721 f/
    27,0 2swap d- 2dup f. ." C - "
    9,0 5,0 f/ f* 32,0 d+ f. ." F" cr
;

: send-temp ( -- )
	begin                                                         
	  key? not                                                    
	while
	  temp 1000 ms                                                                          
	repeat                                                        
	key drop                                                      
;

send-temp

**Reply by tabemann** [reply]
It works, even though I think the CPU temperature sensor is rather inaccurate and jittery.

**Reply by marklacas** [reply]
Just messing about to get my Forth chops back in shape, and to build my zeptoforth knowledge.

**Reply by marklacas** [reply]
Can I use Pad as the buffer for the string conversion?

**Reply by tabemann** [reply]
You *can*, but I would highly recommend `with-allot` or, if cell-alignment is desired, `with-aligned-allot` instead, which have the stack signature ( ? bytes xt -- ? ) and for which xt has the stack signature ( ? addr -- ? ). This is because `pad` easily gets moved and overwritten due to how zeptoforth liberally makes use of the RAM dictionary as a scratch pad, but `with-allot` and `with-aligned-allot` allot space, advancing the `here` address, and afterwards restore the `here` address, even if an exception has taken place (after which it re-raises the exception).

**Reply by marklacas** [reply]
Not really sure how to use with-allot as there is very sparse info about it. Here is my code using pad. It's in the temp word.

```
#include bitmap.fs
#include bitmap_utils.fs
#include font.fs
#include simple_font.fs
#include ssd1306.fs

oo import
bitmap import
ssd1306 import
font import
simple-font import

128 constant my-width
32 constant my-height \ 64 for square display

my-width my-height bitmap-buf-size constant my-buf-size
my-buf-size 4 align buffer: my-buf
<ssd1306> class-size buffer: my-ssd1306
14 15 my-buf my-width my-height SSD1306_I2C_ADDR 1 <ssd1306> my-ssd1306 init-object

init-simple-font

: print-1306 ( string -- )
	my-width 2 / 7 13 * 2 / - my-height 2 / 8 2 / - my-ssd1306 a-simple-font set-string
	$00 my-width 4 / my-width 2 / 0 my-height my-ssd1306 xor-rect-const
	my-ssd1306 update-display
;

adc import

3,3 4096,0 f/ 2constant conversion_factor

\ Send the CPU temperature out to the console
: temp ( -- )
	temp-adc-chan default-adc adc@ s>f conversion_factor f*
	0,706 d-
	2swap 0,001721 f/
	27,0 2swap d- 2dup f. ." C - "
	9,0 5,0 f/ f* 32,0 d+ 2dup f. ." F "
	s"          " print-1306 pad rot rot 1 format-fixed-truncate
	s"  F" 2over + swap dup 2swap rot move + print-1306 cr
;

: send-temp ( -- )
	begin
		key? not
	while
		temp 1000 ms
	repeat
	key drop
;

send-temp

```

**Reply by marklacas** [reply]
It would be nice if the docs had simple examples of usage.

**Reply by tabemann** [reply]
Your code worked for me out of the box (aside from using 64 rather than 32 rows and changing the include paths). However, I would recommend not using `pad` in zeptoforth as, e.g., the SSD1306 driver code uses the RAM dictionary in such a way as to overwrite `pad`. Here is a version of your code, with some revisions, which uses `with-allot` (I have tried out this code as well, with of course the modification to use 64 rows, and changes to the include paths):

```
#include bitmap.fs
#include bitmap_utils.fs
#include font.fs
#include simple_font.fs
#include ssd1306.fs

oo import
bitmap import
ssd1306 import
font import
simple-font import

128 constant my-width
32 constant my-height \ 64 for square display

my-width my-height bitmap-buf-size constant my-buf-size
my-buf-size 4 align buffer: my-buf
<ssd1306> class-size buffer: my-ssd1306
14 15 my-buf my-width my-height SSD1306_I2C_ADDR 1 <ssd1306> my-ssd1306 init-object

init-simple-font

: print-1306 ( string -- )
	my-width 2 / 7 13 * 2 / - my-height 2 / 8 2 / - my-ssd1306 a-simple-font set-string
;

adc import

3,3 4096,0 f/ 2constant conversion_factor

\ Send the CPU temperature out to the console
: temp ( -- )
	temp-adc-chan default-adc adc@ s>f conversion_factor f*
	0,706 d-
	2swap 0,001721 f/
	27,0 2swap d- 2dup f. ." C - "
	9,0 5,0 f/ f* 32,0 d+ 2dup f. ." F "
	my-ssd1306 clear-bitmap
	32 [:
		-rot 1 format-fixed-truncate
        	s"  F" 2over + swap dup 2swap rot move + print-1306
        ;] with-allot
	my-ssd1306 update-display
	cr
;

: send-temp ( -- )
	begin
		key? not
	while
		temp 1000 ms
	repeat
	key drop
;

send-temp
```

Something to note about this code is that it uses a quotation, defined by `[:` and `;]`, which is called by `with-allot`. This is the idiomatic way to use `with-allot`. (You of course can use it with execution tokens gotten through other means, such as `[']`). Note that `with-allot` and `with-aligned-allot` are zeptoforth-specific, so you will not see them in other sources on Forth (unless others have gotten the idea from zeptoforth).

I could spruce up the documentation a good bit, yes, but for the time being I would recommend looking further around the `test` directory in the source tree. (I know there are examples of both `with-allot` and quotations in there.) I have been rather preoccupied though with mitigating the issues with the Wio RP2040's ESP8285 WiFi radio, which have eaten up much of my time as of late.

**Reply by marklacas** [reply]
Thanks, works great.  Good to know about the test code.  I'll dig through there to see what I can find.
